examples/debug.rs

```diff
@@ -0,0 +1,210 @@
+//! Dump the contents of the input archive in a debug format.
+//!
+//! Demonstrates
+//! * iteration through all fundamental types.
+//! * accessing of fields and following of references.
+//!
+//! LICENSE
+//!
+//! The code in this example file is released into the Public Domain.
+
+use osmflat::{Archive, FileResourceStorage, Osm, RefRelationMembers};
+
+use std::fmt;
+use std::str::Utf8Error;
+
+/// Represents fixed point coordinates stored in OSM
+#[derive(Clone, Copy)]
+struct FixedI64(i64);
+
+impl FixedI64 {
+    fn value(&self) -> f64 {
+        self.0 as f64 / osmflat::COORD_SCALE as f64
+    }
+}
+
+impl fmt::Display for FixedI64 {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let value: f64 = self.value();
+        write!(f, "{}", value)
+    }
+}
+
+impl fmt::Debug for FixedI64 {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self)
+    }
+}
+
+#[derive(Debug)]
+struct Header<'ar> {
+    bbox: (FixedI64, FixedI64, FixedI64, FixedI64),
+    required_features: Vec<&'ar str>,
+    optional_features: Vec<&'ar str>,
+    writingprogram: &'ar str,
+    source: &'ar str,
+    osmosis_replication_timestamp: i64,
+    osmosis_replication_sequence_number: i64,
+    osmosis_replication_base_url: &'ar str,
+}
+
+impl<'ar> Header<'ar> {
+    fn new(archive: &'ar Osm) -> Result<Self, Utf8Error> {
+        let header = archive.header();
+        let strings = archive.stringtable();
+
+        let required_features: Result<Vec<_>, _> = (header.required_feature_first_idx() as usize..)
+            .take(header.required_features_size() as usize)
+            .map(|idx| strings.substring(idx))
+            .collect();
+        let optional_features: Result<Vec<_>, _> = (header.optional_feature_first_idx() as usize..)
+            .take(header.optional_features_size() as usize)
+            .map(|idx| strings.substring(idx))
+            .collect();
+
+        Ok(Self {
+            bbox: (
+                FixedI64(header.bbox_left()),
+                FixedI64(header.bbox_right()),
+                FixedI64(header.bbox_top()),
+                FixedI64(header.bbox_bottom()),
+            ),
+            required_features: required_features?,
+            optional_features: optional_features?,
+            writingprogram: strings.substring(header.writingprogram_idx() as usize)?,
+            source: strings.substring(header.source_idx() as usize)?,
+            osmosis_replication_timestamp: header.osmosis_replication_timestamp(),
+            osmosis_replication_sequence_number: header.osmosis_replication_sequence_number(),
+            osmosis_replication_base_url: strings
+                .substring(header.osmosis_replication_base_url_idx() as usize)?,
+        })
+    }
+}
+
+#[derive(Debug)]
```

Comment:
PS: Are all those debug structures, and their conversion function maybe much longer than just printing directly the flatdata structure manually? Brevity might be useful here.

Reply:
Yeah, I am also undecided about that. Somehow printing manually feels weird. Might be useful to learn how to wrap data, but I am not sure.